### ISSUE TITLE
Overlay and Replace are Overly Restrictive

### ISSUE THREAD

**Issue by Measter** [image-rs/image#722]
The type signatures of the `imageops::overlay` and `imageops::replace` functions require that both the bottom and top images be the same type. For example, this code will give a type error:

```rust
let mut bottom  = ImageBuffer::from_pixel(100, 100, Rgba([255, 255, 255, 255]));
let mut top = ImageBuffer::from_pixel(100, 100, Rgba([128, 128, 128, 255]));
let sub_top = top.sub_image(5, 5, 30, 30);
    
image::imageops::overlay(&mut bottom, &sub_top, 5, 5);
```

```
error[E0308]: mismatched types
 --> src/main.rs:9:43
  |
9 |     image::imageops::overlay(&mut bottom, &sub_top, 5, 5);
  |                                           ^^^^^^^^ expected struct `image::ImageBuffer`, found struct `image::SubImage`
  |
  = note: expected type `&image::ImageBuffer<image::Rgba<{integer}>, std::vec::Vec<{integer}>>`
             found type `&image::SubImage<'_, image::ImageBuffer<image::Rgba<{integer}>, std::vec::Vec<{integer}>>>`
```

My proposed fix is to change the signatures to

```rust
pub fn overlay<I: GenericImage, J: GenericImage<Pixel = I::Pixel>>(bottom: &mut I, top: &J, x: u32, y: u32);
pub fn replace<I: GenericImage, J: GenericImage<Pixel = I::Pixel>>(bottom: &mut I, top: &J, x: u32, y: u32);
```

As an aside, couldn't the body of these functions be simplified by zipping together two ranges for the coordinates, like this:

```rust
for (by, ty) in (y..bottom_height).zip(0..top_height) {
    for (bx, tx) in (x..bottom_width).zip(0..top_width) {
       // ...
    }
}
```

That would handle bounds-checking because a range with a start higher than the end returns `None` (e.g. `y` >= `bottom_height`), and the zip will end when the shortest range finishes.

**Comment by HeroicKatora** [comment]
Wow, I like that bounds checking even more than #850 .. And it is much simpler.

**Comment by HeroicKatora** [comment]
Fixed in #880
